onnxruntime/core/session/symbols.txt

```diff
@@ -0,0 +1,12 @@
+# we provide stubs for these functions in the ORT C API implementation so that we return a graceful error message 
```

Comment:
This would add additional exports (that were not necessarily built with) to the ORT DLL. Another option would be to add the non-struct overloads of these functions to the C API and have the C# binding call them using SessionOptionsAppendExecutionProvider_*. This way users don't have to include the additional provider_factory headers for using individual EPs and will pave the way towards having all EP APIs in one place rather than scattered around.

Overall, the setup (not this PR) is super confusing. Some EPs are implemented as DLLs, some not. APIs for some EPs are declared only in the corresponding factory header and for some they're declared in both c_api.h and the corresponding factory header (for e.g. CUDA, TensorRT, OpenVINO). For CUDA, the same function is declared in both the places.

Reply:
Agree it's extremely confusing. Maybe the best time to clean all this up is when all the EPs are separate libraries so that side of the setup is consistent. At that point we should get rid of all the standalone functions and require everything to go via the API. 

I considered adding things to the C API but doing so doesn't resolve the lack of symbols for existing functions that breaks an iOS build. I would have to remove the existing functions and add new equivalent functions to the C API, however that would break any existing code and probably should be done in two steps.

Reply:
We don't have any plans to make the other EPs separate libs in the near or distant future. But yeah cleaning up the APIs is indeed a bigger change. We can address it in a separate PR. Added AB#1334627 to track it.